### PR TITLE
a few changes which help out GC performance

### PR DESCRIPTION
**Pull request by toshok**
A few changes (mostly dealing with scanning) that help GC perf.

```
              pyston (calibration)                      :    1.0s baseline: 1.0 (-0.4%)
              pyston django_migrate.py                  :    5.9s baseline: 10.7 (-45.1%)
              pyston interp2.py                         :    6.0s baseline: 5.9 (+1.5%)
              pyston raytrace.py                        :    6.7s baseline: 6.8 (-1.1%)
              pyston nbody.py                           :    8.9s baseline: 9.3 (-4.0%)
              pyston fannkuch.py                        :    8.1s baseline: 8.0 (+0.8%)
              pyston chaos.py                           :   21.3s baseline: 20.5 (+4.2%)
              pyston fasta.py                           :    6.0s baseline: 5.6 (+8.2%)
              pyston pidigits.py                        :    5.4s baseline: 5.5 (-0.6%)
              pyston richards.py                        :    2.0s baseline: 2.0 (-0.9%)
              pyston deltablue.py                       :    2.1s baseline: 2.1 (+0.8%)
              pyston (geomean-8999)                     :    5.9s baseline: 6.2 (-5.0%)
```

### CODE REVIEW

File: src/codegen/ast_interpreter.cpp [hunk 1]
```diff
@@ -69,7 +69,8 @@ union Value {
 
 class ASTInterpreter {
 public:
-    typedef llvm::DenseMap<InternedString, Box*> SymMap;
+    typedef llvm::DenseMap<InternedString, int> SymMap;
+    typedef std::vector<Box*> SymVector;
```
Comment (kmod) [inline src/codegen/ast_interpreter.cpp]: Maybe we should factor this out a separate datastructure?  ex a "ContiguousMap" (or whatever) that abstracts these operations.

File: src/codegen/ast_interpreter.cpp [hunk 2]
```diff
@@ -165,7 +168,8 @@ class ASTInterpreter {
 void ASTInterpreter::addSymbol(InternedString name, Box* value, bool allow_duplicates) {
     if (!allow_duplicates)
         assert(sym_table.count(name) == 0);
-    sym_table[name] = value;
+    sym_table[name] = sym_vector.size();
+    sym_vector.push_back(value);
```
Comment (kmod) [inline src/codegen/ast_interpreter.cpp]: what if name was already in the sym_table?

File: src/codegen/ast_interpreter.cpp (hunk 1 above)
Comment (toshok) [inline src/codegen/ast_interpreter.cpp]: Yeah, I added ContiguousMap, and then found a few places where I hadn't converted over, so it was a good change.

File: src/core/contiguous_map.h [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+    const_iterator begin() const noexcept { return map.begin(); }
+    const_iterator end() const noexcept { return map.end(); }
+
+    iterator erase(const_iterator position) {
+        int idx = map[position->first];
+        free_list.push_back(idx);
+        vec[idx] = TVal();
+        map.erase(position->first);
+        return begin(); // this is broken...
```
Comment (kmod) [inline src/core/contiguous_map.h]: Could this return the return value of map.erase?  Also, are we guaranteed that the provided iterator is in the map?

**Pull request merged**